src/cargo/sources/overlay.rs

```diff
@@ -0,0 +1,119 @@
+use std::task::ready;
+
+use crate::sources::IndexSummary;
+
+use super::source::{MaybePackage, Source};
+
+/// A `Source` that overlays a local registry over a remote registry, pretending
+/// that the packages in the local registry are actually available in the remote
+/// registry.
+///
+/// This is a massive footgun and a terrible idea, so we do not (and never will)
+/// expose this publicly. However, it is useful for some very specific private
+/// things, like locally verifying a bunch of packages at a time before any of
+/// them have been published.
+pub struct OverlayConfusionAttack<'gctx> {
+    local: Box<dyn Source + 'gctx>,
+    remote: Box<dyn Source + 'gctx>,
+}
+
+impl<'gctx> OverlayConfusionAttack<'gctx> {
+    pub fn new(local: Box<dyn Source + 'gctx>, remote: Box<dyn Source + 'gctx>) -> Self {
+        Self { local, remote }
+    }
+}
+
+impl<'gctx> Source for OverlayConfusionAttack<'gctx> {
+    fn source_id(&self) -> crate::core::SourceId {
+        self.remote.source_id()
+    }
+
+    fn supports_checksums(&self) -> bool {
+        true
+    }
+
+    fn requires_precise(&self) -> bool {
+        false
+    }
+
+    fn query(
+        &mut self,
+        dep: &crate::core::Dependency,
+        kind: super::source::QueryKind,
+        f: &mut dyn FnMut(super::IndexSummary),
+    ) -> std::task::Poll<crate::CargoResult<()>> {
+        let local_source = self.local.source_id();
+        let remote_source = self.remote.source_id();
+
+        let local_dep = dep.clone().map_source(remote_source, local_source);
+        let mut callback = |index: IndexSummary| {
+            let index = index.map_summary(|s| s.map_source(local_source, remote_source));
+            f(index)
+        };
+        ready!(self.local.query(&local_dep, kind, &mut callback))?;
+        self.remote.query(dep, kind, f)
```

Comment:
Does this need to handle the same package/version being available from both registries? Possibly given its limited use case, just by detecting the situation and crashing.

Reply:
Good point, I've added that check and a test.